### PR TITLE
Vlasiator-4.0

### PR DESCRIPTION
**Pull request by ursg**
This pull requests supersedes the previous pull requests #336 (Multipop), #329 (Marconi support) and #319 (Fsgrid), which are merged together and cleanly updated to current master here. It also encompasses issuse #230 and #332.

It has been used for production runs on Marconi (both AED and Yann's shock tests).

Consolidating the TODO lists remaining from the earlier requests, there are still some open ends to be tied here:

  - [x] Update testpackage to do the same tests as before, and add multipop test
  - [x] Split off the remaining boundary options to be population-wise.
  - [x] Split off backstream parameters ? (Discussion below)
  - [x] Fix fsgrid/cellparams Hall term E components IO output. (See below.)
 - [x] Test and merge https://github.com/ursg/vlasiator/pull/19
 - [x] Review all data reduction operators (At least TWO reviewers!). See https://github.com/ursg/vlasiator/pull/21 for details.
 - [ ] Review all data reduction operators (At least TWO reviewers!).

As soon as these are resolved, and current production runs with the old codebase are through, I suggest merging this into master.

### CODE REVIEW

**Comment by ykempf** [comment]
I suppose the backstream parameters could/should also be defined population-wise, for the sake of some generality.

**Comment by ykempf** [comment]
I spotted a small bug for IO. Hall E is not being pulled out of fsgrid to cellparams for IO. I will put that into a commit later/ping me about it if I forget. The fix includes Hall E and GradPe E if they are to be written out.

**Comment by ykempf** [comment]
The later commit I also need to compare the new version with a reference test I was running right now.

**Comment by ykempf** [comment]
Whoops there was a wrong parenthesis closure there. Glad I tested and caught it.

I now confirm that the change related to the Hall term in the two commits c80869d and a173ad1 does not modify the E given out and it yields EHall components if the output is set.

**Comment by ykempf** [comment]
I'll take care of the backstreaming parameter splitting.

**Comment by ykempf** [comment]
`ERROR in vlsv::ParallelReader! I only got -32766/0 bytes in vlsv_reader_parallel.cpp:519
ERROR, failed to read BLOCKVARIABLE in ioread.cpp:349
ERROR, failed to read BLOCKIDS in ioread.cpp:345`

Excerpt from Marconi output files.

**Comment by ykempf** [comment]
Also:
slurmstepd: error: Detected 1 oom-kill event(s) in step 586688.1 cgroup.
srun: error: r106c02s02: tasks 5351,5358: Out Of Memory
srun: Terminating job step 586688.1
slurmstepd: error: Detected 1 oom-kill event(s) in step 586688.1 cgroup.
slurmstepd: error: *** STEP 586688.1 ON r066c11s04 CANCELLED AT 2018-03-31T23:03:11 ***

But this happens after the logfile said we're exiting.

**Comment by ykempf** [comment]
https://github.com/fmihpc/vlsv/pull/33 Will fix the former (of the last 2 reported errors).

**Comment by ykempf** [comment]
@ursg, please note the latest commit, I had the population backstream options mistakenly with an underscore instead of a full stop.

File: sysboundary/outflow.cpp [hunk 1]
```diff
@@ -127,6 +89,78 @@ namespace SBC {
...
+        }
+        if(!Readparameters::get(pop + "_outflow.vlasovScheme_face_z+", vlasovSysBoundarySchemeName[4])) {
+           if(myRank == MASTER_RANK) cerr << __FILE__ << ":" << __LINE__ << " ERROR: This option has not been added for population " << pop << "!" << endl;
+           exit(1);
+        }
+        if(!Readparameters::get(pop + "_outflow.vlasovScheme_face_z-", vlasovSysBoundarySchemeName[5])) {
+           if(myRank == MASTER_RANK) cerr << __FILE__ << ":" << __LINE__ << " ERROR: This option has not been added for population " << pop << "!" << endl;
+           exit(1);
+        }
+        for(uint i=0; i<6 ; i++) {
```
Comment (ykempf) [inline sysboundary/outflow.cpp]: @ursg I don't like embedded for loops with the same index... Even though you told me it should work all right...

**Comment by ykempf** [comment]
The current <population>/PTensorDiagonal and <population>/PTensorOffDiagonal have the same data as the DRO does a loop over all species. Either this is stored as the global PTensor and does not need to be duplicated per species, or this is supposed to be the partial pressure (and the summation can be done later if needed).

Opinions?

**Comment by markusbattarbee** [comment]
This should definitely be per-species, so we can analyse T_parallel and T_perpendicular per-species.

**Comment by ykempf** [comment]
Here it is.

**Comment by ursg** [comment]
With all the recend feature branches merged into this one, and all production runs of Vlasiator 3.0 complete, I think it is time to merge this into master. The conflcts below already have a recorded preimage in the vlasiator-4.0-preview branch, so they will present no problem.

**Pull request merged**